pkg/rc/replication_controller.go

```diff
@@ -0,0 +1,243 @@
+package rc
+
+import (
+	"strings"
+	"time"
+
+	"github.com/square/p2/pkg/kp"
+	"github.com/square/p2/pkg/kp/rcstore"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/pods"
+	"github.com/square/p2/pkg/rc/fields"
+	"github.com/square/p2/pkg/util"
+)
+
+const (
+	rcIdLabel  = "replication_controller_id"
+	podIdLabel = "pod_id"
+)
+
+// These methods are the same as the methods of the same name in kp.Store.
+// Replication controllers have no need of any methods other than these.
+type kpStore interface {
+	SetPod(key string, manifest pods.Manifest) (time.Duration, error)
+	DeletePod(key string) (time.Duration, error)
+}
+
+type replicationController struct {
+	fields.RC
+
+	logger logging.Logger
+
+	kpStore       kpStore
+	rcStore       rcstore.Store
+	scheduler     Scheduler
+	podApplicator labels.Applicator
+}
+
+func New(
+	fields fields.RC,
+	kpStore kpStore,
+	rcStore rcstore.Store,
+	scheduler Scheduler,
+	podApplicator labels.Applicator,
+) *replicationController {
+	return &replicationController{
+		RC: fields,
+
+		logger:        logging.DefaultLogger,
+		kpStore:       kpStore,
+		rcStore:       rcStore,
+		scheduler:     scheduler,
+		podApplicator: podApplicator,
+	}
+}
+
+func (rc *replicationController) WatchDesires(quit <-chan struct{}) <-chan error {
+	desiresChanged, errInChannel := rc.rcStore.Watch(&rc.RC, quit)
+
+	errOutChannel := make(chan error)
+	channelsClosed := make(chan struct{})
+
+	// When seeing any changes, try to meet then.
+	// If meeting produces any error, send it on the output error channel.
+	go func() {
+		for _ = range desiresChanged {
+			err := rc.meetDesires()
+			if err != nil {
+				errOutChannel <- err
+			}
+		}
+		channelsClosed <- struct{}{}
+	}()
+
+	// When seeing any errors, forward them to the output error channel.
+	// (Notice that two goroutines may be writing to the output error channel)
+	go func() {
+		for err := range errInChannel {
+			errOutChannel <- err
+		}
+		channelsClosed <- struct{}{}
+	}()
+
+	// If both channels I'm watching are closed,
+	// which would happen when my caller requests a quit
+	// (the quit gets forwarded to the rc watcher)
+	// close my output channel.
+	go func() {
+		<-channelsClosed
+		<-channelsClosed
+		close(channelsClosed)
+		close(errOutChannel)
+	}()
+
+	return errOutChannel
+}
+
+func (rc *replicationController) meetDesires() error {
+	rc.logger.NoFields().Infof("Meeting with desired replicas %d, disabled %v", rc.ReplicasDesired, rc.Disabled)
+
+	// If we're disabled, we do nothing, nor is it an error
+	// (it's a normal possibility to be disabled)
+	if rc.Disabled {
+		return nil
+	}
+
+	eligible, err := rc.eligibleNodes()
```

Comment:
yeah makes sense. moved it into the only branch where it's necessary then. and removed the separate log line anyway since there's another log line that covers it